File: bindings/matlab/benchmarks/bench_pqc.m
% libpqc-dyber - Post-Quantum Cryptography Library
% Copyright (c) 2024-2026 Dyber, Inc.
% SPDX-License-Identifier: Apache-2.0 OR MIT
%
% MATLAB binding benchmarks for all PQC algorithms.
% Uses tic/toc for timing with statistical reporting.

function bench_pqc(base_iters)
    if nargin < 1
        base_iters = 100;
    end

    DEFAULT_ITERATIONS = base_iters;
    SLOW_ITERATIONS = 5;
    WARMUP_ITERATIONS = 5;

    pqc_init();

    fprintf('language,algorithm,operation,iterations,');
    fprintf('min_ms,max_ms,mean_ms,median_ms,stddev_ms,ops_per_sec,');
    fprintf('pk_bytes,sk_bytes\n');

    % KEM benchmarks
    kem_names = pqc_kem_algorithm_names();
    for k = 1:length(kem_names)
        name = kem_names{k};
        iters = adjusted_iters(name, DEFAULT_ITERATIONS, SLOW_ITERATIONS);
        kem = pqc_kem_new(name);
        pk_size = pqc_kem_public_key_size(kem);
        sk_size = pqc_kem_secret_key_size(kem);

        % Warmup
        for w = 1:WARMUP_ITERATIONS
            pqc_kem_keygen(kem);
        end

        % Keygen
        samples = zeros(1, iters);
        for i = 1:iters
            tic;
            pqc_kem_keygen(kem);
            samples(i) = toc * 1000;
        end
        csv_row('matlab', name, 'keygen', iters, compute_stats(samples), pk_size, sk_size);

        % Encaps
        [pk, sk] = pqc_kem_keygen(kem);
        for i = 1:iters
            tic;
            pqc_kem_encaps(kem, pk);
            samples(i) = toc * 1000;
        end
        csv_row('matlab', name, 'encaps', iters, compute_stats(samples), pk_size, sk_size);

        % Decaps
        [ct, ss] = pqc_kem_encaps(kem, pk);
        for i = 1:iters
            tic;
            pqc_kem_decaps(kem, ct, sk);
            samples(i) = toc * 1000;
        end
        csv_row('matlab', name, 'decaps', iters, compute_stats(samples), pk_size, sk_size);

        pqc_kem_free(kem);
    end

    % Signature benchmarks
    msg = uint8(mod((0:1023) * 137 + 42, 256));

    sig_names = pqc_sig_algorithm_names();
    for k = 1:length(sig_names)
        name = sig_names{k};
        iters = adjusted_iters(name, DEFAULT_ITERATIONS, SLOW_ITERATIONS);
        sig = pqc_sig_new(name);
        pk_size = pqc_sig_public_key_size(sig);
        sk_size = pqc_sig_secret_key_size(sig);
        if pqc_sig_is_stateful(sig)
            iters = min(iters, SLOW_ITERATIONS);
        end

        % Keygen
        samples = zeros(1, iters);
        for i = 1:iters
            tic;
            pqc_sig_keygen(sig);
            samples(i) = toc * 1000;
        end
        csv_row('matlab', name, 'keygen', iters, compute_stats(samples), pk_size, sk_size);

        % Sign
        [pk, sk] = pqc_sig_keygen(sig);
        for i = 1:iters
            tic;
            pqc_sig_sign(sig, msg, sk);
            samples(i) = toc * 1000;
        end
        csv_row('matlab', name, 'sign(1KB)', iters, compute_stats(samples), pk_size, sk_size);

        % Verify
        signature = pqc_sig_sign(sig, msg, sk);
        for i = 1:iters
            tic;
            pqc_sig_verify(sig, msg, signature, pk);
            samples(i) = toc * 1000;
        end
        csv_row('matlab', name, 'verify(1KB)', iters, compute_stats(samples), pk_size, sk_size);

        pqc_sig_free(sig);
    end

    pqc_cleanup();
end

function iters = adjusted_iters(name, base, slow_max)
    if contains(name, 'McEliece') || contains(name, 'Frodo') || ...
       contains(name, 'XMSS') || contains(name, 'LMS')
        iters = min(base, slow_max);
    else
        iters = base;
    end
end

function s = compute_stats(samples)
    sorted = sort(samples);
    n = length(sorted);
    s.min_v = sorted(1);
    s.max_v = sorted(end);
    s.median_v = median(sorted);
    s.mean_v = mean(sorted);
    if n > 1
        s.stddev_v = std(sorted);
    else
        s.stddev_v = 0;
    end
    if s.mean_v > 0
        s.ops = 1000.0 / s.mean_v;
    else
        s.ops = 0;
    end
end

function csv_row(lang, algo, op, iters, s, pk_size, sk_size)
    fprintf('%s,%s,%s,%d,%.6f,%.6f,%.6f,%.6f,%.6f,%.1f,%d,%d\n', ...
        lang, algo, op, iters, ...
        s.min_v, s.max_v, s.mean_v, s.median_v, ...
        s.stddev_v, s.ops, pk_size, sk_size);
end

File: bindings/rust/benches/bench_pqc.rs
// libpqc-dyber - Post-Quantum Cryptography Library
// Copyright (c) 2024-2026 Dyber, Inc.
// SPDX-License-Identifier: Apache-2.0 OR MIT
//
// Rust binding benchmarks for all PQC algorithms.
// Uses criterion for statistical benchmarking of keygen, encaps/sign,
// decaps/verify across every enabled algorithm.

use criterion::{criterion_group, criterion_main, Criterion, BenchmarkId};
use pqc_dyber::{kem, sig};

fn bench_kem_algorithms(c: &mut Criterion) {
    let mut group = c.benchmark_group("KEM");

    for alg_name in kem::algorithm_names() {
        let kem_ctx = match kem::Kem::new(alg_name) {
            Ok(k) => k,
            Err(_) => continue,
        };

        group.bench_with_input(
            BenchmarkId::new("keygen", alg_name),
            alg_name,
            |b, _| {
                b.iter(|| kem_ctx.keygen().unwrap());
            },
        );

        let (pk, sk) = kem_ctx.keygen().unwrap();

        group.bench_with_input(
            BenchmarkId::new("encaps", alg_name),
            alg_name,
            |b, _| {
                b.iter(|| kem_ctx.encaps(&pk).unwrap());
            },
        );

        let (ct, _ss) = kem_ctx.encaps(&pk).unwrap();

        group.bench_with_input(
            BenchmarkId::new("decaps", alg_name),
            alg_name,
            |b, _| {
                b.iter(|| kem_ctx.decaps(&ct, &sk).unwrap());
            },
        );
    }

    group.finish();
}

fn bench_sig_algorithms(c: &mut Criterion) {
    let mut group = c.benchmark_group("Signature");
    let msg = vec![0x42u8; 1024]; // 1KB test message

    for alg_name in sig::algorithm_names() {
        let sig_ctx = match sig::Signature::new(alg_name) {
            Ok(s) => s,
            Err(_) => continue,
        };

        // Skip stateful algorithms in criterion (they need special handling)
        if sig_ctx.is_stateful() {
            continue;
        }

        group.bench_with_input(
            BenchmarkId::new("keygen", alg_name),
            alg_name,
            |b, _| {
                b.iter(|| sig_ctx.keygen().unwrap());
            },
        );

        let (pk, sk) = sig_ctx.keygen().unwrap();

        group.bench_with_input(
            BenchmarkId::new("sign/1KB", alg_name),
            alg_name,
            |b, _| {
                b.iter(|| sig_ctx.sign(&msg, &sk).unwrap());
            },
        );

        let signature = sig_ctx.sign(&msg, &sk).unwrap();

        group.bench_with_input(
            BenchmarkId::new("verify/1KB", alg_name),
            alg_name,
            |b, _| {
                b.iter(|| sig_ctx.verify(&msg, &signature, &pk).unwrap());
            },
        );
    }

    group.finish();
}

criterion_group!(benches, bench_kem_algorithms, bench_sig_algorithms);
criterion_main!(benches);
